src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c

```diff
@@ -0,0 +1,483 @@
+
+#include "pa_unix_util.h"
+#include "pa_ringbuffer.h"
+
+#include "pa_hostapi_pulseaudio_cb.h"
+
+
+/* Pulseaudio headers */
+#include <stdio.h>
+#include <string.h>
+#include <pulse/pulseaudio.h>
+
+/* prototypes for functions declared in this file */
+
+#define PA_PULSEAUDIO_SET_LAST_HOST_ERROR(errorCode, errorText) \
+    PaUtil_SetLastHostErrorInfo(paInDevelopment, errorCode, errorText)
+
+
+
+void PulseaudioStreamReadCb(pa_stream *s, size_t length, void *userdata)
+{
+    PaPulseaudioStream *l_ptrStream = (PaPulseaudioStream *) userdata;
+    PaStreamCallbackTimeInfo timeInfo = {0, 0, 0}; /* TODO: IMPLEMENT ME */
+    int l_iResult = paContinue;
+    long numFrames = 0;
+    int i = 0;
+    size_t l_lDataSize = 0;
+    size_t l_lLocation = 0;
+    size_t l_lBufferSize = 0;
+    const void *l_ptrSampleData = NULL;
+
+    assert(s);
+    assert(length > 0);
+
+    if(l_ptrStream == NULL)
+    {
+        PA_DEBUG(("Portaudio %s: Stream is null!\n", __FUNCTION__));
+        return;
+    }
+
+    memset(l_ptrStream->inBuffer, 0x00, PULSEAUDIO_BUFFER_SIZE);
+
+
+    PaUtil_BeginCpuLoadMeasurement(&l_ptrStream->cpuLoadMeasurer);
+
+    while (pa_stream_readable_size(s) > 0)
+    {
+        l_ptrSampleData = NULL;
+
+        if(pa_stream_peek(s, &l_ptrSampleData, &l_lDataSize))
+        {
+            PA_DEBUG(("Portaudio %s: Can't read audio!\n", __FUNCTION__));
+            return;
+
+        }
+
+        if (l_lDataSize > 0 &&  l_ptrSampleData)
+        {
+            memcpy(l_ptrStream->inBuffer + l_lBufferSize, l_ptrSampleData, l_lDataSize);
+            l_lBufferSize += l_lDataSize;
+        }
+        else
+        {
+            PA_DEBUG(("Portaudio %s: Can't read!\n", __FUNCTION__));
+        }
+
+
+        pa_stream_drop(s);
+
+    }
+
+    if(l_ptrStream->bufferProcessor.streamCallback != NULL)
+    {
+        PaUtil_BeginBufferProcessing(&l_ptrStream->bufferProcessor, &timeInfo, 0);
+        PaUtil_SetInterleavedInputChannels(&l_ptrStream->bufferProcessor, 0, l_ptrStream->inBuffer, l_ptrStream->inSampleSpec.channels);
+        PaUtil_SetInputFrameCount(&l_ptrStream->bufferProcessor, l_lBufferSize / l_ptrStream->inputFrameSize);
+        numFrames = PaUtil_EndBufferProcessing(&l_ptrStream->bufferProcessor, &l_iResult);
+
+        PaUtil_EndCpuLoadMeasurement(&l_ptrStream->cpuLoadMeasurer, numFrames);
+    }
+    else
+    {
+        PaUtil_WriteRingBuffer( &l_ptrStream->inputRing, l_ptrStream->inBuffer, l_lBufferSize );
+    }
+
+    if(l_iResult != paContinue)
+    {
+        l_ptrStream->isActive = 0;
+        return;
+    }
+
+
+    pa_threaded_mainloop_signal(l_ptrStream->mainloop, 0);
+}
+
+void PulseaudioStreamWriteCb(pa_stream *s, size_t length, void *userdata)
+{
+    PaPulseaudioStream *l_ptrStream = (PaPulseaudioStream *) userdata;
+    PaStreamCallbackTimeInfo timeInfo = {0, 0, 0}; /* TODO: IMPLEMENT ME */
+    int l_iResult = paContinue;
+    long numFrames = 0;
+    int i = 0;
+
+    assert(s);
+    assert(length > 0);
```

Comment:
Ah you are right, this is the way Portaudio is designed. So keep it, we have to only be sure, that it catches only coding issues and does not crash because of any runtime exception on the users machine.